mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,478 @@
+    derived_variables : Dict[str, DerivedVariable]
+        Dictionary with the variables to derive with keys as the variable
+        names and values with entries for kwargs and function to use in
+        the calculation
+    chunking: Dict[str, int]
+        Dictionary with keys as the dimensions to chunk along and values
+        with the chunk size
+
+    Returns
+    -------
+    xr.Dataset
+        Dataset with derived variables included
+    """
+
+    ds_derived_vars = xr.Dataset()
+    ds_derived_vars.attrs.update(ds.attrs)
+    for _, derived_variable in derived_variables.items():
+        required_kwargs = derived_variable.kwargs
+        function_name = derived_variable.function
+        derived_variable_attributes = derived_variable.attributes
+
+        # Separate the lat,lon from the required variables as these will be derived separately
+        logger.warning(
+            "Assuming that the lat/lon coordinates are given as variables called"
+            " 'lat' and 'lon'."
+        )
+        latlon_coords_to_include = {}
+        for key in list(required_kwargs.keys()):
+            if key in ["lat", "lon"]:
+                latlon_coords_to_include[key] = required_kwargs.pop(key)
+
+        # Get input dataset for calculating derived variables
+        ds_input = ds[required_kwargs.keys()]
+
+        # Any coordinates needed for the derivation, for which chunking should be performed,
+        # should be converted to variables since it is not possible for *indexed* coordinates
+        # to be chunked dask arrays
+        chunks = {
+            dim: chunking.get(dim, int(ds_input[dim].count())) for dim in ds_input.dims
+        }
+        required_coordinates = [
+            req_var for req_var in required_kwargs.keys() if req_var in ds_input.coords
+        ]
+        ds_input = ds_input.drop_indexes(required_coordinates, errors="ignore")
+        for req_coord in required_coordinates:
+            if req_coord in chunks:
+                ds_input = ds_input.reset_coords(req_coord)
+
+        # Chunk the dataset
+        ds_input = _chunk_dataset(ds_input, chunks)
+
+        # Add function arguments to kwargs
+        kwargs = {}
+        if len(latlon_coords_to_include):
+            latlon = get_latlon_coords_for_input(ds)
+            for key, val in latlon_coords_to_include.items():
+                kwargs[val] = latlon[key]
+        kwargs.update({val: ds_input[key] for key, val in required_kwargs.items()})
+        func = _get_derived_variable_function(function_name)
+        # Calculate the derived variable
+        derived_field = func(**kwargs)
+
+        # Check the derived field(s) and add it to the dataset
+        if isinstance(derived_field, xr.DataArray):
+            derived_field = _check_attributes(
+                derived_field, derived_variable_attributes
+            )
+            ds_derived_vars[derived_field.name] = derived_field
+        elif isinstance(derived_field, tuple) and all(
```

Comment:
The functions that returns two data-arrays currently are the ones where we drive day of year and hour of day. Since I have implemented a cyclic encoding in them, the cosine and sine part are both returned. I thought that it wouldn't make any sense to return only e.g. hour of day as a non-cyclically encoded variable and that is why I did it like this. Not sure if we wanted to split them up how that should be implemented then. The other option is then to not do the cyclic encoding here.